sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala

```diff
@@ -55,8 +55,45 @@ import org.apache.spark.util.{ParentClassLoader, Utils}
  *                 to null.
  * @param value A term for a (possibly primitive) value of the result of the evaluation. Not
  *              valid if `isNull` is set to `true`.
+ * @param inputRow A term that holds the input row name when generating this code.
+ * @param inputVars A list of [[ExprInputVar]] that holds input variables when generating this code.
  */
-case class ExprCode(var code: String, var isNull: String, var value: String)
+case class ExprCode(
+    var code: String,
+    var isNull: String,
+    var value: String,
+    var inputRow: String = null,
+    var inputVars: Seq[ExprInputVar] = Seq.empty) {
+
+  // Returns true if this value is a literal.
+  def isLiteral(): Boolean = {
+    assert(value.nonEmpty, "ExprCode.value can't be empty string.")
+
+    if (value == "true" || value == "false" || value == "null") {
+      true
+    } else {
+      // The valid characters for the first character of a Java variable is [a-zA-Z_$].
+      value.head match {
+        case v if v >= 'a' && v <= 'z' => false
```

Comment:
This won't work for array accessing like `arr[1]`, I think we should pick a more general solution like https://github.com/apache/spark/pull/19938#issuecomment-350935431 , i.e. generate parameter names instead of reusing the input name.

Reply:
Is array accessing like `arr[1]` all global variables or not?

Reply:
hmmm this seems very hard to do, the code is already generated and use the input names as whatever it is, e.g. java variable `a` or literal `123` or array accessing `arr[1]`. Ideally we need to analyze what the input really refers, e.g. `a` refers to a java variable `a`, `123` refers nothing, `arr[1]` refers to a java variable `arr`. This is kind of impossible for the current string based framework. We need to think more about how to deal with it.

Reply:
also cc @kiszk @mgaido91 @maropu

Reply:
I only consider literals and normal java variables now. If we say array accessing `arr[1]` are all global variables (I suppose they are, as usually we don't/can't use `arr[1]` as variable name), I'm thinking maybe we should disallow passing any global variables.

Reply:
can we do a regex check to make sure `value` is a valid java variable name? We can make an assumption here that if it's not a valid java variable name, it should be accessible globally, like `123`, or `arr[1]` while `arr` is a global variable.

Reply:
another idea, let's move `isLiteral` and `isEvaluated` to `ExpressionCodegen`. Seems these 2 are not general definition of `ExprCode`, and only makes sense for `ctx.currentVars`.

Reply:
It is doable. Instead of `isLiteral`, we may call it `isValidJavaVariable`.

Reply:
or `isGloballyAccess`?

Reply:
> another idea, let's move isLiteral and isEvaluated to ExpressionCodegen. Seems these 2 are not general definition of ExprCode, and only makes sense for ctx.currentVars.

Ok.

Reply:
thanks! sorry for the trouble...

Reply:
no problem at all.